wandb/sdk/launch/_project_spec.py

```diff
@@ -247,6 +243,29 @@ def _fetch_job(self) -> None:
         job.configure_launch_project(self)
         self._job_artifact = job._job_artifact
 
+    def get_image_source_string(self) -> str:
+        """Returns a unique string identifying the source of an image"""
+        if self.source == LaunchSource.LOCAL:
+            # TODO: more correct to get a hash of local uri contents
+            assert isinstance(self.uri, str)
+            return self.uri
+        elif self.source == LaunchSource.JOB:
+            assert isinstance(self.job, str)
+            return self.job
+        elif self.source == LaunchSource.GIT:
+            assert isinstance(self.uri, str)
+            ret = self.uri
+            if self.git_version:
+                ret += self.git_version
+            return ret
+        elif self.source == LaunchSource.WANDB:
+            assert isinstance(self.uri, str)
+            return self.uri
+        elif self.source == LaunchSource.DOCKER:
+            raise LaunchError("Attempted to get image source string for docker image")
```

Comment:
Is this an actionable message for a user? We could _logger.debug this, but im not sure I would know what to do as a user if I saw this error.

Reply:
This function is used for tagging the image with a unique identifier. We will always build an image unless the user specifies a job with a unique artifact index. (see `check_build_required`).